File: minimum_depth_binary_tree.py
In DFS/recursive implementation you may end up going down one sub-tree and explore all the nodes.
In BFS there are chances we find solution faster as we do level order traversal and return depth of first leaf node.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root: return 0
        queue = []
        queue.append({'n':root, 'd':1})
        
        while len(queue)>0:
            ele = queue.pop(0)
            node, depth = ele['n'], ele['d']
            if(node):
                if node.left is None and node.right is None:
                    return depth
                if node.left is not None:
                    queue.append({'n':node.left, 'd':depth + 1})
                if node.right is not None:
                    queue.append({'n':node.right, 'd':depth + 1})

File: minimum_depth_binary_tree.cpp
 /**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    //recursive
    int minDepth(TreeNode* root) {
        if (!root) return 0;
        if (!root->left and !root->right) return 1; 
        
        if (!root->left) 
            return minDepth(root->right) + 1;
        
        if (!root->right) 
            return minDepth(root->left) + 1;
        
        return min(minDepth(root->left), minDepth(root->right)) + 1;
    }
    
    //iterative
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        
        queue<TreeNode*> qu;
        TreeNode *current = root;
        qu.push(current);
        
        int mindepth = 1;
        
        while(!qu.empty()) {
            
            int n = qu.size(); 
            while(n-- > 0) {
                current = qu.front();
                qu.pop();
            
                if(!current->left and !current->right) return mindepth;
                
                if(current->left) qu.push(current->left);
                if(current->right) qu.push(current->right);
                    
            }
            mindepth++;
        }
        return mindepth;
    }
};
